src/name/ip_address.rs

```diff
@@ -14,6 +14,236 @@
 
 use crate::Error;
 
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+
+const VALID_IP_BY_CONSTRUCTION: &str = "IP address is a valid string by construction";
+
+/// Either a Ipv4 or Ipv6 address, plus its owned string representation
+#[cfg(feature = "alloc")]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub enum IpAddress {
+    /// An ipv4 address and its owned string representation
+    IpV4Address(String, [u8; 4]),
+    /// An ipv6 address and its owned string representation
+    IpV6Address(String, [u8; 16]),
+}
+
+#[cfg(feature = "alloc")]
+impl AsRef<str> for IpAddress {
+    fn as_ref(&self) -> &str {
+        match self {
+            IpAddress::IpV4Address(ip_address, _) | IpAddress::IpV6Address(ip_address, _) => {
+                ip_address.as_str()
+            }
+        }
+    }
+}
+
+/// Either a ipv4 or ipv6 address, plus its borrowed string representation
+#[derive(Debug, Clone, Copy)]
+pub enum IpAddressRef<'a> {
+    /// An ipv4 address and its borrowed string representation
+    IpV4AddressRef(&'a [u8], [u8; 4]),
```

Comment:
Given that these byte slices are known to hold valid UTF-8, why not just use `&str` here?

Reply:
I agree. Going to check it.

Reply:
I just had the time to revisit this. I was trying to be as close as possible to the current code on webpki. In this case, in DNS names, although they are also known to hold valid UTF-8, [`DNSNameRef` also refers to `&[u8]`](https://github.com/briansmith/webpki/blob/b481381e914de604df30b1be2fe72d938a5c0e57/src/name/dns_name.rs#L80).

Given the goal was to contribute this to webpki upstream I tried to deviate as less as possible from the existing codebase.

Reply:
Okay, fair enough.